### PR TITLE
Add resource ID to all data resources

### PR DESCRIPTION
**Pull request by uri-weisman**
## What does this PR do?
Adds a resource id to all of our data resources.

The resource id value for each of the data resources will be:
processes - process id.
files - inode uid.
kube-api - metadata uid.

## Why is it important?
Will provide us the ability to use elastic's transform feature and present data according to a specific resource.

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Author's Checklist

- [ ]

## Related issues

-

### CODE REVIEW

File: cloudbeat/constructor/processor_test.go [hunk 1]
```diff
@@ -0,0 +1,109 @@
+package constructor
```
Comment (uri-weisman) [inline cloudbeat/constructor/processor_test.go:1]: Please ignore for now

File: cloudbeat/constructor/processor.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+	}
+}
+
+func (c *Constructor) processEachResource(ctx context.Context, client beat.Client, results []fetchers.FetcherResult, metadata ResourceTypeMetadata) {
+	for _, result := range results {
+		events, err := c.createBeatEvents(ctx, result, ResourceMetadata{ResourceTypeMetadata: metadata, ResourceId: result.Resource.GetID()})
+		if err != nil {
+			fmt.Errorf("failed to create beat events for, %v, Error: %v", metadata, err)
+		}
+		client.PublishAll(events)
```
Comment (uri-weisman) [inline cloudbeat/constructor/processor.go:41]: I consider using channels to publish the events on the cloudbeat business layer.

File: cloudbeat/beater/cloudbeat.go [hunk 3]
```diff
@@ -62,27 +62,27 @@ func New(b *beat.Beat, cfg *common.Config) (beat.Beater, error) {
...
+
 	bt := &cloudbeat{
 		ctx:          ctx,
 		cancel:       cancel,
 		config:       c,
 		eval:         evaluator,
 		data:         data,
-		resultParser: eventParser,
-		scheduler:    scheduler,
+		resultsIndex: resultsIndex,
```
Comment (eyalkraft) [inline cloudbeat/beater/cloudbeat.go:84]: note it's not used

File: cloudbeat/beater/cloudbeat.go [hunk 4]
```diff
@@ -152,16 +147,16 @@ func InitRegistry(ctx context.Context, c config.Config) (resources.FetchersRegis
...
 	}
 
 	procCfg := fetchers.ProcessFetcherConfig{
 		Directory: processesDir,
 	}
-	procf := fetchers.NewProcessesFetcher(procCfg)
 
-	if err = registry.Register("processes", procf); err != nil {
+	procf := fetchers.NewProcessesFetcher(procCfg)
+	if err = registry.Register(fetchers.ProcessType, procf); err != nil {
```
Comment (eyalkraft) [inline cloudbeat/beater/cloudbeat.go:159]: same change for "file_system" fetcher is missing?

File: cloudbeat/resources/fetchers/file_system_fetcher_test.go [hunk 5]
```diff
@@ -28,9 +28,12 @@ func TestFileFetcherFetchASingleFile(t *testing.T) {
 	assert.Nil(t, err, "Fetcher was not able to fetch files from FS")
```
Comment (uri-weisman) [inline cloudbeat/resources/fetchers/file_system_fetcher_test.go:28]: Plz ignore for now

File: cloudbeat/constructor/processor.go (hunk 2 above)
Comment (eyalkraft) [reply cloudbeat/constructor/processor.go:41]: Yeah I'm not very fond of passing the client all the way as well.
A channel could be perfect but maybe in another PR.
let's not pass client through all this chain of functions and return the events to kubebeat.go for now

File: cloudbeat/beater/cloudbeat.go [hunk 6]
```diff
@@ -115,17 +115,12 @@ func (bt *cloudbeat) Run(b *beat.Beat) error {
...
 			bt.updateCycleStatus(cycleId, cycleStatusStart)
-
-			resourceCallback := func(resource interface{}) {
-				bt.resourceIteration(bt.ctx, resource, cycleId)
-			}
-
-			bt.scheduler.ScheduleResources(o, resourceCallback)
-
+			cycleMetadata := constructor.CycleMetadata{CycleId: cycleId}
+			bt.constructor.ProcessAggregatedResources(bt.ctx, bt.client, fetchedResources, cycleMetadata)
```
Comment (eyalkraft) [inline cloudbeat/beater/cloudbeat.go:123]: `ctx` can move to NewConstructor, `bt.client` shouldn't be passed (see previous comment)

**Review by eyalkraft — COMMENTED** [review]
short draft review before sending to more team members

**Comment by uri-weisman** [comment]
While implementing this logic - adding a `resource_id` to our beat events, @eyalkraft and I noticed that we lose the resource context in the scheduler (resource type for example), and a new param needs to be passed all the way to the parser to be used, which feels like something that will make us a hard time in the future.
As a mid-time solution, we thought about creating a BL that will handle the results parsing & beat event creation.

@amirbenun, @oren-zohar - as this is not a minor change would appreciate your thoughts on this draft PR.

File: cloudbeat/resources/data.go [hunk 7]
```diff
@@ -147,15 +147,14 @@ func copyState(m Map) (Map, error) {
 
 func init() {
 	gob.Register([]interface{}{})
-	gob.Register(FetcherResult{})
```
Comment (amirbenun) [inline cloudbeat/resources/data.go:150]: Are you sure we can drop it?

File: cloudbeat/config/config_test.go [hunk 8]
```diff
@@ -1,3 +1,4 @@
+//go:build !integration
```
Comment (amirbenun) [inline cloudbeat/config/config_test.go:1]: What does it mean?

File: cloudbeat/constructor/constructor.go [hunk 9]
```diff
@@ -0,0 +1,26 @@
+package constructor
```
Comment (amirbenun) [inline cloudbeat/constructor/constructor.go:1]: naming is confusing

File: cloudbeat/constructor/processor.go [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+}
+
+func (c *Constructor) createBeatEvents(resource interface{}, metadata ResourceMetadata) error {
+	result, err := c.callback(c.context, resource)
+	if err != nil {
+		logp.Error(fmt.Errorf("error running the policy: %w", err))
+		return err
+	}
+
+	findings, err := ParseResult(result)
```
Comment (amirbenun) [inline cloudbeat/constructor/processor.go:62]: I would expect to see the deletion of `evaluation_result_parser.go`

**Review by amirbenun — COMMENTED** [review]
Feels like a step in the right direction

File: cloudbeat/beater/cloudbeat.go (hunk 3 above)
Comment (uri-weisman) [reply cloudbeat/beater/cloudbeat.go:83]: Removed

File: cloudbeat/resources/data.go (hunk 7 above)
Comment (uri-weisman) [reply cloudbeat/resources/data.go:150]: Yes sir

**Review by eyalkraft — APPROVED** [review]

File: cloudbeat/resources/data.go [hunk 11]
```diff
@@ -4,40 +4,40 @@ import (
...
-	"github.com/elastic/beats/v7/libbeat/common/kubernetes"
 	"github.com/elastic/beats/v7/libbeat/logp"
 )
 
 // Data maintains a cache that is updated by Fetcher implementations registered
 // against it. It sends the cache to an output channel at the defined interval.
 type Data struct {
 	interval time.Duration
-	output   chan Map
+	output   chan ResourceMap
```
Comment (oren-zohar) [inline cloudbeat/resources/data.go:19]: That was the original name and we changed it to `Map` for some reason I can't recall, @amirbenun @yashtewari do you have any idea?

**Pull request closed without merging**